From 82783551b318e7bebbe7a51add2aa22388a9f2bb Mon Sep 17 00:00:00 2001
From: Junxiao Bi <junxiao.bi@windriver.com>
Date: Wed, 14 Mar 2012 18:07:56 +0800
Subject: [PATCH] octeon: net: fix wrong tx packets free in synchronous IO mode

The counter(skb_to_free) that records how many skb should be freed
isn't handled well in synchronous IO mode. This will cause some packets
freed even before hardware finishs sending them.

Signed-off-by: Junxiao Bi <junxiao.bi@windriver.com>
---
 drivers/net/cavium-ethernet/ethernet-tx.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/cavium-ethernet/ethernet-tx.c b/drivers/net/cavium-ethernet/ethernet-tx.c
index 5cc4825..97fb4f3 100644
--- a/drivers/net/cavium-ethernet/ethernet-tx.c
+++ b/drivers/net/cavium-ethernet/ethernet-tx.c
@@ -280,7 +280,7 @@ dont_put_skbuff_in_hw:
 		buffers_to_free = cvmx_scratch_read64(CVMX_SCR_SCRATCH+8);
 	} else {
 		/* Get the number of skbuffs in use by the hardware */
-		skb_to_free = cvmx_fau_fetch_and_add32(priv->fau+qos*4, 1);
+		skb_to_free = cvmx_fau_fetch_and_add32(priv->fau+qos*4, MAX_SKB_TO_FREE);
 		buffers_to_free = cvmx_fau_fetch_and_add32(FAU_NUM_PACKET_BUFFERS_TO_FREE, 0);
 	}
 
-- 
1.7.0.2

